test/langtools/tools/javac/processing/filer/TestNoOverwriteJarFiles.java

```diff
@@ -0,0 +1,291 @@
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8338675
+ * @summary javac shouldn't silently change .jar files on the classpath
+ * @library /tools/lib /tools/javac/lib
+ * @modules jdk.compiler/com.sun.tools.javac.api
+ *          jdk.compiler/com.sun.tools.javac.main
+ * @build toolbox.ToolBox toolbox.JarTask toolbox.JavacTask
+ * @run junit TestNoOverwriteJarFiles
+ */
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.parallel.Execution;
+import toolbox.JavacTask;
+import toolbox.ToolBox;
+
+import javax.annotation.processing.RoundEnvironment;
+import javax.lang.model.element.TypeElement;
+import javax.tools.FileObject;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileTime;
+import java.time.Instant;
+import java.util.Set;
+import java.util.jar.JarOutputStream;
+import java.util.zip.ZipEntry;
+
+import static javax.tools.StandardLocation.CLASS_OUTPUT;
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.parallel.ExecutionMode.SAME_THREAD;
+
+/**
+ * Tests that javac cannot unexpectedly modify contents of JAR files on the
+ * class path.
+ * <p>
+ * Consider the following javac behaviours:
+ * <ol>
+ * <li>If there is no source path, javac searches the classpath for sources,
+ * including inside JAR files.
+ * <li>If a Java source file was modified more recently that an existing class
+ * file, or if no class file exists, javac will compile it from the source.
+ * <li>If there is no output directory specified, javac will put compiled class
+ * files next to their corresponding sources.
+ * </ol>
+ * Taken together, this suggests that a newly compiled class file should be
+ * written back into the JAR in which its source was found, possibly overwriting
+ * an existing class file entry. This would be very problematic.
+ * <p>
+ * This test ensures javac will not modify JAR files on the classpath, even if
+ * it compiles sources contained within them. Instead, the class file will be
+ * written into the current working directory, which mimics the JDK 8 behavior.
+ *
+ * <h2>Important</h2>
+ *
+ * This test creates files from Java compilation and annotation processing, and
```

Comment:
I suppose some of the desired effect could be achieved using jtreg directives instead, but this approach seems fine too.